### PR TITLE
sam: fix HD line parsing for cases where tags are user defined

### PR DESCRIPTION
**Pull request by kortschak**
I ended up going with the othertags approach. The publicly visible map turned out to be a lot more complicated.

@brentp Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
If you think this is OK, I'll extend the logic to SQ, RG and PG lines.

**Comment by brentp** [comment]
LGTM

**Comment by kortschak** [comment]
Added logic and tests for other types.

Fixes #26.

PTAL

File: sam/header.go [hunk 1]
```diff
@@ -126,12 +134,85 @@ func NewHeader(text []byte, r []*Reference) (*Header, error) {
...
+}
+
 // Clone returns a deep copy of the receiver.
 func (bh *Header) Clone() *Header {
 	c := &Header{
 		Version:    bh.Version,
 		SortOrder:  bh.SortOrder,
 		GroupOrder: bh.GroupOrder,
+		otherTags:  append([]tagPair(nil), bh.otherTags...),
 		Comments:   append([]string(nil), bh.Comments...),
```
Comment (brentp) [inline sam/header.go]: I haven't seen this idiom, it just allows you to create (and append to) an empty slice in-place?

**Comment by brentp** [comment]
LGTM

File: sam/header.go (hunk 1 above)
Comment (kortschak) [inline sam/header.go]: Use it sparingly. It's the equivalent of `func(src []T) []T { dst := make([]T, len(src)); copy(dst, src); return dst }`. rsc has expressed significant distaste for it, but here it makes sense where it's in a struct declaration and there are only two cases - with different types.

**Pull request merged**